3p/ampcontext.js

```diff
@@ -171,12 +171,17 @@ export class AmpContext {
     // to check the name attribute as it has been bypassed.
     // TODO(alanorozco): why the heck could AMP_CONTEXT_DATA be two different
     // types? FIX THIS.
-    if (!this.win_.AMP_CONTEXT_DATA) {
-      this.setupMetadata_(this.win_.name);
-    } else if (typeof this.win_.AMP_CONTEXT_DATA == 'string') {
-      this.sentinel = this.win_.AMP_CONTEXT_DATA;
+    if (/** @type {?Object} */(this.win_.sf_) &&
+        /** @type {?Object} */(this.win_.sf_.cfg)) {
+      this.setupMetadata_(this.win_.sf_.cfg);
+    } else if (this.win_.AMP_CONTEXT_DATA) {
+      if (typeof this.win_.AMP_CONTEXT_DATA == 'string') {
+        this.sentinel = this.win_.AMP_CONTEXT_DATA;
+      } else {
+        this.setupMetadata_(this.win_.AMP_CONTEXT_DATA);
+      }
     } else {
-      this.setupMetadata_(this.win_.AMP_CONTEXT_DATA);
+      this.setupMetadata_(this.win_.name);
```

Comment:
Couple of questions about setupMetadata_:

- It may throw if name cannot be JSON parsed?  Where does that error get caught?  What is the side effect?
- It assumes the object given matches the context object attributes (e.g. has field named pageViewId).  Can you instead cast to a context object?